### PR TITLE
[SYCL][PM] Fix DeadArgumentEliminationSYCLPass with new pass manager

### PR DESCRIPTION
**Pull request by pvchupin**
* Add it properly
* Add test to check presence in optimized compilations
* Fix documentation, it's enabled by default long time ago

### CODE REVIEW

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 1]
```diff
@@ -1301,6 +1301,9 @@ PassBuilder::buildModuleOptimizationPipeline(OptimizationLevel Level,
   if (!LTOPreLink)
     MPM.addPass(RelLookupTableConverterPass());
 
+  if (SYCLOptimizationMode)
+    MPM.addPass(DeadArgumentEliminationSYCLPass());
+
```
Comment (aelovikov-intel) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:1306]: I'm not sure if this is the right place or somewhere in https://github.com/intel/llvm/blob/273034ab6920add3330627765ac53f9d31f6c5cd/clang/lib/CodeGen/BackendUtil.cpp#L1491-L1506

Comment (pvchupin) [reply llvm/lib/Passes/PassBuilderPipelines.cpp:1306]: I had the same doubts actually... On one hand it seems this pass belongs to optimization pipeline... on the other this seems to be the best location to call device-only pass. In latter case I had a problem writing the clang test, cannot see the pass in clang -cc1 output...

Comment (asudarsa) [reply llvm/lib/Passes/PassBuilderPipelines.cpp]: I am not sure is 'SYCLOptimizationMode' is available inside clang pipeline. Current location looks good to me.

Comment (asudarsa) [reply llvm/lib/Passes/PassBuilderPipelines.cpp]: I just had a look at the changes in clang. The current testing looks good. However, do we want to have testing to verify if the dead arguments (if any) are actually getting eliminated?

thanks

File: clang/test/CodeGenSYCL/dead-elim-sycl.cpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// RUN: %clang -cc1 -triple spir64-unknown-unknown  -fsycl-is-device  -emit-llvm-bc  -O1 -fno-legacy-pass-manager -fdebug-pass-manager -fenable-sycl-dae -o /dev/null -x c++ < %s 2>&1 | FileCheck %s
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/dead-elim-sycl.cpp:1]: How about adding one more run line that tests with -fdisable-llvm-passes?

File: llvm/lib/Passes/PassBuilderPipelines.cpp (hunk 1 above)
Comment (aelovikov-intel) [reply llvm/lib/Passes/PassBuilderPipelines.cpp]: I think that should be covered by either opt LIT test (I assume one must exist already) or a full blown end-to-end test, not here and probably outside of this PR's scope.

File: clang/test/CodeGenSYCL/dead-elim-sycl.cpp (hunk 2 above)
Comment (aelovikov-intel) [reply clang/test/CodeGenSYCL/dead-elim-sycl.cpp:1]: Done

**Review by premanandrao — APPROVED** [review]
FE changes LGTM

**Review by smanna12 — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]

**Review by asudarsa — APPROVED** [review]
LGTM. Thanks.

File: llvm/lib/Passes/PassBuilderPipelines.cpp (hunk 1 above)
Comment (pvchupin) [reply llvm/lib/Passes/PassBuilderPipelines.cpp]: There is some test at: llvm/test/Transforms/DeadArgElim/sycl-kernels-neg.ll

**Comment by aelovikov-intel** [comment]
@pvchupin , your approval is needed on behalf of @intel/dpcpp-doc-reviewers .

**Comment by pvchupin** [comment]
> @pvchupin , your approval is needed on behalf of @intel/dpcpp-doc-reviewers .
Github thinks I'm still the author, so doesn't allow me self-approve, can merge without it as soon as testing pass.
... and we need to add somebody else to doc reviewers to back me up

**Pull request merged**